wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java

```diff
@@ -485,4 +492,84 @@ public String toString(E value) {
       this.value = value;
     }
   }
+
+  private static final class MapProtoAdapter<K, V> extends ProtoAdapter<Map<K, V>> {
+    private final MapEntryProtoAdapter<K, V> entryAdapter;
+
+    MapProtoAdapter(ProtoAdapter<K> keyAdapter, ProtoAdapter<V> valueAdapter) {
+      super(FieldEncoding.LENGTH_DELIMITED, null);
+      entryAdapter = new MapEntryProtoAdapter<>(keyAdapter, valueAdapter);
+    }
+
+    @Override public int encodedSize(Map<K, V> value) {
+      throw new UnsupportedOperationException("Repeated values can only be sized with a tag.");
+    }
+
+    @Override public int encodedSizeWithTag(int tag, Map<K, V> value) {
+      int size = 0;
+      for (Map.Entry<K, V> entry : value.entrySet()) {
+        size += entryAdapter.encodedSizeWithTag(tag, entry);
+      }
+      return size;
+    }
+
+    @Override public void encode(ProtoWriter writer, Map<K, V> value) {
+      throw new UnsupportedOperationException("Repeated values can only be encoded with a tag.");
+    }
+
+    @Override public void encodeWithTag(ProtoWriter writer, int tag, Map<K, V> value)
+        throws IOException {
+      for (Map.Entry<K, V> entry : value.entrySet()) {
+        entryAdapter.encodeWithTag(writer, tag, entry);
+      }
+    }
+
+    @Override public Map<K, V> decode(ProtoReader reader) throws IOException {
```

Comment:
I can add on `newMapAdapter` but it's also true for `asRepeated()` and `List`. Because of how adapters work we can only ever decode one element at a time. In the case of `List` we delegate directly to the item adapter and skip the singleton list. Unfortunately because of the complex type here we can't do that.